keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/resource/ClusterHandler.kt

```diff
@@ -789,79 +732,54 @@ class ClusterHandler(
     atZone(ZoneId.systemDefault()).format(DateTimeFormatter.ISO_DATE_TIME)
 
   /**
-   * Translates a LaunchConfiguration object to a ClusterSpec.LaunchConfigurationSpec with default values omitted for export.
+   * Translates a ServerGroup object to a ClusterSpec.ServerGroupSpec with default values omitted for export.
    */
-  private fun LaunchConfiguration.exportSpec(account: String, application: String): ClusterSpec.LaunchConfigurationSpec {
-    val defaultKeyPair = cloudDriverCache.defaultKeyPairForAccount(account)
-    return ClusterSpec.LaunchConfigurationSpec(
-      instanceType = instanceType,
-      ebsOptimized = if (!ebsOptimized) {
-        null
-      } else {
-        ebsOptimized
-      },
-      // for the iam role, we compare against a default based on a stable naming convention
-      iamRole = if (iamRole == LaunchConfiguration.defaultIamRoleFor(application)) {
-        null
-      } else {
-        iamRole
-      },
-      // for the key pair, we compare against the currently configured default in clouddriver since there are
-      // multiple naming conventions, and handle the case where the default includes a region placeholder
-      keyPair = if (defaultKeyPair.contains(REGION_PLACEHOLDER)) {
-        if (defaultKeyPair.replace(REGION_PLACEHOLDER, REGION_PATTERN).toRegex().matches(keyPair)) {
-          null
-        } else {
-          keyPair
-        }
-      } else {
-        if (keyPair == defaultKeyPair) {
-          null
-        } else {
-          keyPair
-        }
-      },
-      instanceMonitoring = if (!instanceMonitoring) {
+  private fun ServerGroup.exportSpec(account: String, application: String): ClusterSpec.ServerGroupSpec {
+    val defaults = ClusterSpec.ServerGroupSpec(
```

Comment:
Unfortunately, all the spec types generally have all nullable fields initalized with `null`, as opposed to default values as commonly found in some of the clouddriver models. There's a `resolve()` method (not part of any interface) generally implemented in the top-level object like `ClusterSpec` which actually resolves several of these defaults, but in terms of the clouddriver models, which also doesn't help here. We could definitely refactor that code to be more reusable, but for now I think this is a good-enough compromise.

Reply:
We either implement it like that (null defaults) because we need you to specify it _either_ cluster-wide _or_ in the region specific overrides. I don't see a way to do that with the type system alone. If we assign defaults on object construction we would not be able to say later whether that default was something intentionally set by the user, or just a default.

For example, should the default values found in region specific overrides trump the non-default value found cluster wide? What if the user had actually specified the default value in one region, but now _all_ the regions have the default because of the way the constructor works?

Reply:
Yeah, I know... it’s complicated, which is why I gave up on refactoring the types. The only thing that occurred to me is that we could have yet another set of data classes for overrides so we could maintain proper non-nullability in the class that represents defaults, but it seemed like a lot of work for little benefit.

The other problem is that, beyond hard-coded defaults in the class constructors, we have “dynamic defaults” calculated by resolvers — those are the hardest ones to account for in the export flow as far as omitting fields with default values. I cheated a bit with things like the key pair stuff because that resolver is in our code base, but the moment someone adds a resolver as an extension, there’s no way I could think of to also dynamically figure out what to omit. Not with the current APIs anyway.